salt/returners/sqlalchemy.py

```diff
@@ -0,0 +1,314 @@
+
+    Requires that configuration be enabled via 'event_return'
+    option in master config.
+    """
+    Events = model_for("Events", engine_name=__context__.get("engine_name"))
+    with salt.sqlalchemy.Session(__context__.get("engine_name")) as session:
+        records = []
+        for evt in evts:
+            record = {
+                "tag": evt.get("tag", ""),
+                "data": evt.get("data", ""),
+                "cluster": __opts__["cluster_id"],
+                "master_id": __opts__["id"],
+            }
+
+            try:
+                record["created_at"] = evt["data"]["_stamp"]
+            except (KeyError, TypeError):
+                pass
+
+            records.append(record)
+
+        session.execute(insert(Events), records)
+        session.commit()
+
+
+@retry_on_failure
+def save_load(jid, load, minions=None):
+    """
+    Save the load to the specified jid id
+    """
+    if not minions:
+        minions = []
+
+    Jids = model_for("Jids", engine_name=__context__.get("engine_name"))
+    with salt.sqlalchemy.Session(__context__.get("engine_name")) as session:
+        record = {
+            "jid": jid,
+            "load": load,
+            "minions": minions,
+            "cluster": __opts__["cluster_id"],
+        }
+
+        session.execute(insert(Jids), [record])
+        session.commit()
+
+
+def save_minions(jid, minions, syndic_id=None):  # pylint: disable=unused-argument
+    """
+    Included for API consistency
+    """
+
+
+def get_load(jid):
+    """
+    Return the load data that marks a specified jid
+    """
+    Jids = model_for("Jids", engine_name=__context__.get("engine_name"))
+    with salt.sqlalchemy.ROSession(__context__.get("engine_name")) as session:
+        # Use to_jsonb for jsonb conversion in Postgres
+        stmt = select(Jids).where(Jids.jid == str(jid))
+        result = session.execute(stmt).first()
+        load = {}
+        if result:
+            jid = result[0]
+            load = jid.load
+            load["Minions"] = jid.minions or []
+        session.commit()
+        return load
+
+
+def get_jid(jid):
+    """
+    Return the information returned when the specified job id was executed
+    """
+    Returns = model_for("Returns", engine_name=__context__.get("engine_name"))
+    with salt.sqlalchemy.ROSession(__context__.get("engine_name")) as session:
+        stmt = select(Returns.id, Returns.ret).where(Returns.jid == str(jid))
+        results = session.execute(stmt).all()
+
+        ret = {}
+        for result in results:
+            ret[result.id] = result.ret
+
+        session.commit()
+
+        return ret
+
+
+def get_fun(fun):
+    """
+    Return a dict of the last function called for all minions
+    """
+    # this could be done with a separate table, but why?
+    raise salt.exceptions.SaltException(
+        "This is too costly to run via salt_returns, left unimplemented"
+    )
+
+
+def get_jids(last=None):
```

Comment:
Executing `salt-run jobs.list_jobs` requires the `get_jids` function. If SQLAlchemy is used as the cache, does that mean this command can no longer be used?

Reply:
hmm yes. I can look at adding it back in some form possibly. the problem is if you have a deploy of any meaningful size, running salt-run jobs.list_jobs without any args is 1 letter away from list_job, except it OOM's your master; hence me shortcircuiting it. there might be a way to make it work safely

Reply:
Perhaps it may be worth exploring if adding pagination controls in the configuration layer could improve system responsiveness? For example:

```yaml
sqlalchemy:
  max_query_rows_per_page: 1000  # Max rows per query page
  max_query_pages: 10            # Maximum page fetches 
  max_returner_items: 1000        # Salt returner specific (defaults to max_query_rows_per_page)
```

Would this approach help enforce consistent pagination behavior across SQLAlchemy operations? It might reduce memory pressure and improve response efficiency for both databases and the Salt Master - potentially preventing oversized result sets from impacting performance.

Reply:
looking closer at this, really the filtering interface/kwargs need to be moved into get_jids() itself, so the database can do the work. I think this can be tackled as a followup PR/feature work though, as this PR is already pretty large